rust_src/src/keymap.rs

```diff
@@ -317,6 +319,96 @@ pub fn map_keymap_lisp(function: LispObject, keymap: LispObject, sort_first: boo
     LispObject::constant_nil()
 }
 
+/// Call FUN for every binding in MAP and stop at (and return) the parent.
+/// FUN is called with 4 arguments: FUN (KEY, BINDING, ARGS, DATA).  */
+#[no_mangle]
+pub extern "C" fn map_keymap_internal(
+    map: Lisp_Object,
+    fun: map_keymap_function_t,
+    args: Lisp_Object,
+    data: *const c_void,
+) -> Lisp_Object {
+    let mut tail = LispObject::constant_nil();
+    let map = LispObject::from_raw(map);
```

Comment:
Instead of defining tail and then updating it you could assign here.

```
let mut tail = match map.as_cons() {
    None => LispObject::constant_nil(),
    Some(cons) => {
        let (car, cdr) = cons.as_tuple();
        if car.eq_raw(Qkeymap) {
            cdr
        } else {
            map
        }
};
````

Reply:
I see. I should get the hang of this kind of format soon. ;)